File: web/apps/web/src/channels/use-hls-player/device-profile/device-profile.ts
import type { DeviceProfileBody } from "@loomarr/api/models/deviceProfileBody";

// deviceProfile() — what THIS browser can direct-play, as a DeviceProfile the play-url endpoint uses
// (§9.1 V48/V50). Its role is a yes/no GATE, not a plan picker: under the V50 content-driven model the
// CHANNEL's broadcast codec decides the timeline codec/container, and this profile only tells the
// server whether this client can decode that codec natively. When the channel is HEVC and the browser
// can decode HEVC, the server serves a `-c:v copy` fMP4 stream instead of a full HEVC→h264 transcode
// (the black-frame-free, GPU-free path a media server has always had); when it can't, the server
// down-converts that same channel to h264/TS for this client. An h264 channel is served h264/TS to
// everyone regardless of what this profile advertises.
//
// ⚠ We ask MediaSource.isTypeSupported, NOT video.canPlayType — the same lesson the transport split
// learned (use-hls-player.ts): Chromium's canPlayType lies ("maybe" for things it cannot decode),
// while MSE (which is the pipeline hls.js actually feeds) answers for the real decoder. h264/aac are
// the implied floor and never sent; we only report what we can prove ABOVE it, so a browser that
// proves nothing gets the safe baseline. Being conservative here is correct: a false positive is a
// black frame, a false negative is only a transcode.

// isTypeSupported is a thin, guarded wrapper — MediaSource may be absent (old Safari uses native HLS
// and never constructs one), in which case we advertise nothing and the server serves baseline.
const isTypeSupported = (mime: string): boolean => {
  const MS = typeof self !== "undefined" ? self.MediaSource : undefined;
  return !!MS && typeof MS.isTypeSupported === "function" && MS.isTypeSupported(mime);
};

// HEVC codec strings. Main profile (hvc1.1.6.L93.B0) is the 8-bit baseline; Main 10 (hvc1.2.4.L120.B0)
// is the 10-bit profile. We test both because they gate DIFFERENT plans server-side (hevc8 vs the
// 10-bit part of hevc10), and a decoder can support one without the other. `hev1`/`hvc1` are the two
// sample-entry fourccs; a decoder that does HEVC accepts at least one, so we OR them.
const HEVC_8BIT = ['video/mp4; codecs="hvc1.1.6.L93.B0"', 'video/mp4; codecs="hev1.1.6.L93.B0"'];
const HEVC_10BIT = ['video/mp4; codecs="hvc1.2.4.L120.B0"', 'video/mp4; codecs="hev1.2.4.L120.B0"'];
// Surround audio the browser can decode in-MSE. EAC3/AC3 support is real on Safari and some Chromium
// builds; where absent, the server transcodes the audio to AAC (video still copies). mp4a.a6 = EAC3,
// mp4a.a5 = AC3.
const EAC3 = 'audio/mp4; codecs="mp4a.a6"';
const AC3 = 'audio/mp4; codecs="mp4a.a5"';

const anySupported = (mimes: string[]): boolean => mimes.some(isTypeSupported);

// deviceProfile builds the body sent on POST /v1/channels/{id}/play-url. Only capabilities ABOVE the
// h264/aac floor are listed; an empty result (no HEVC, no surround) is a valid profile that resolves
// to baseline. maxResolution comes from the screen so the server can cap the ladder for a small device.
const deviceProfile = (): DeviceProfileBody => {
  const video: string[] = [];
  const audio: string[] = [];

  const hevc8 = anySupported(HEVC_8BIT);
  const hevc10 = anySupported(HEVC_10BIT);
  if (hevc8 || hevc10) video.push("hevc");
  if (isTypeSupported(EAC3)) audio.push("eac3");
  if (isTypeSupported(AC3)) audio.push("ac3");

  const px = Math.round(Math.max(window.screen.width, window.screen.height) * (window.devicePixelRatio || 1));

  return {
    video,
    audio,
    // 10-bit is only claimed when the 10-bit HEVC profile itself decodes — the server gates its
    // 10-bit copy on this exact bit, so a decoder that does 8-bit HEVC only must NOT set it.
    video10bit: hevc10,
    // HDR display is not something MSE exposes reliably; leave it to a future, explicit probe rather
    // than guess. Omitted (false) keeps the server on the SDR/tone-map-safe path.
    hdr: false,
    maxResolution: Number.isFinite(px) && px > 0 ? px : 0,
  };
};

export { deviceProfile };

File: android/app/src/main/java/loomarr/media/playback/DeviceProfile.kt
package loomarr.media.playback

import android.media.MediaCodecInfo
import android.media.MediaCodecList
import org.json.JSONArray
import org.json.JSONObject

/**
 * What this device can actually decode, probed at runtime.
 *
 * ⚠ PROBED, never hardcoded. The same APK runs on a Shield Pro that decodes HEVC 10-bit and on a
 * cheap stick that manages h264 only, and the server bases its copy-vs-transcode decision on what
 * we claim here. A false positive is a black screen; a false negative is only a needless transcode,
 * so every check below fails toward the safe answer.
 *
 * This mirrors how Jellyfin's Android TV client builds its profile — from `MediaCodecList` at
 * runtime, posted per playback — rather than shipping a device table.
 */
data class DeviceProfile(
    val video: List<String>,
    val audio: List<String>,
    val video10Bit: Boolean,
    val hdr: Boolean,
    val maxResolution: Int,
) {
    fun toJson(): JSONObject =
        JSONObject()
            .put("video", JSONArray(video))
            .put("audio", JSONArray(audio))
            .put("video10bit", video10Bit)
            .put("hdr", hdr)
            .put("maxResolution", maxResolution)

    companion object {
        /**
         * Probe the device.
         *
         * ⚠ Software decoders are EXCLUDED. A software HEVC decoder exists on most devices and will
         * technically decode, but at 4K it drops frames badly — advertising it turns a "we can play
         * this" promise into a stuttering picture. Jellyfin filters these out for the same reason;
         * the codec being *listed* is not the same as it being *usable*.
         */
        fun probe(): DeviceProfile {
            val codecs = MediaCodecList(MediaCodecList.REGULAR_CODECS)
            val video = mutableSetOf<String>()
            val audio = mutableSetOf<String>()
            var tenBit = false
            var hdr = false

            // ⚠ Starts UNKNOWN, not at 1080.
            //
            // Seeded to 1080 and combined with maxOf, this could never report anything BELOW 1080:
            // a device whose HEVC decoder tops out at 720 still claimed 1080, because the seed
            // won. A floor dressed as a measurement — the probe exists precisely so capability is
            // read from the device rather than assumed, and this assumed the common answer.
            //
            // Harmless today only because the server does not yet read `maxResolution` (it is
            // plumbed to playout.CopyPlan and never consumed). The day it gates a copy plan, this
            // would hand low-end hardware a stream it cannot decode, and the symptom — a black
            // screen — would surface nowhere near this line.
            var maxHeight: Int? = null

            for (info in codecs.codecInfos) {
                if (info.isEncoder) continue
                if (isSoftwareOnly(info)) continue

                for (type in info.supportedTypes) {
                    when (type.lowercase()) {
                        "video/avc" -> video += "h264"
                        "video/hevc" -> {
                            video += "hevc"
                            val caps = runCatching { info.getCapabilitiesForType(type) }.getOrNull()
                            if (caps != null) {
                                if (supportsHevc10Bit(caps)) tenBit = true
                                if (supportsHdr(caps)) hdr = true
                                maxHeight = tallerOf(maxHeight, maxHeightFor(caps))
                            }
                        }
                        // AV1 is probed like anything else rather than special-cased. The Shield has
                        // no AV1 decoder and never will; a newer box does, and the probe is what
                        // tells them apart without a device table.
                        "video/av01" -> video += "av1"
                        "audio/mp4a-latm" -> audio += "aac"
                        "audio/ac3" -> audio += "ac3"
                        "audio/eac3" -> audio += "eac3"
                    }
                }
            }

            // h264/aac are the implied floor the server assumes anyway; including them keeps the
            // posted profile self-describing rather than relying on that assumption.
            video += "h264"
            audio += "aac"

            return DeviceProfile(
                video = video.toList(),
                audio = audio.toList(),
                video10Bit = tenBit,
                hdr = hdr,
                // ⚠ 0 means "no cap", which is the wire contract's own word for unknown — the
                // server documents `0 = no cap` and omitempty drops the field entirely. So a device
                // that would not report a height says nothing rather than guessing, and the server
                // falls back to whatever it does for a client that never sent a profile.
                //
                // Deliberately NOT a conservative low guess either: inventing 720 would cap a
                // device that never claimed to be limited, transcoding streams it could have
                // direct-played. Silence is the honest answer to a question the device declined.
                maxResolution = maxHeight ?: 0,
            )
        }

        /**
         * A codec is software-only when the platform says so, or when its name carries one of the
         * conventional software prefixes.
         *
         * ⚠ `isSoftwareOnly` is API 29+, and `minSdk` is 23, so the version guard is load-bearing
         * rather than defensive — calling it unguarded would crash on an older device. The name
         * prefixes are the fallback below 29 and still catch oddly-named vendor codecs above it.
         */
        private fun isSoftwareOnly(info: MediaCodecInfo): Boolean {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q && info.isSoftwareOnly) {
                return true
            }
            val name = info.name.lowercase()
            return name.startsWith("omx.google.") || name.startsWith("c2.android.")
        }

        // ⚠ The HDR profile CONSTANTS are themselves API-gated — Main10HDR10 is API 24, HDR10Plus is
        // API 29 — so they are compared as literals rather than referenced by name. Referencing the
        // constant on an older device throws NoSuchFieldError at class-load time, which is a crash
        // rather than a missing capability. Their integer values are frozen platform API and cannot
        // change; a device too old to define them simply never reports them.
        // Values read from android.jar (API 36) with `javap -constants`, NOT from memory — a wrong
        // literal here silently reports the wrong capability, which is a black screen rather than a
        // compile error. HevcProfileConstantsTest pins them.
        private const val HEVC_PROFILE_MAIN10 = 2
        private const val HEVC_PROFILE_MAIN10_HDR10 = 4096
        private const val HEVC_PROFILE_MAIN10_HDR10_PLUS = 8192

        private fun supportsHevc10Bit(caps: MediaCodecInfo.CodecCapabilities): Boolean =
            caps.profileLevels.any {
                it.profile == HEVC_PROFILE_MAIN10 ||
                    it.profile == HEVC_PROFILE_MAIN10_HDR10 ||
                    it.profile == HEVC_PROFILE_MAIN10_HDR10_PLUS
            }

        private fun supportsHdr(caps: MediaCodecInfo.CodecCapabilities): Boolean =
            caps.profileLevels.any {
                it.profile == HEVC_PROFILE_MAIN10_HDR10 || it.profile == HEVC_PROFILE_MAIN10_HDR10_PLUS
            }

        /**
         * The taller of two reported heights, where null means "not reported".
         *
         * A device may expose several hardware HEVC decoders, and what it can play is the best of
         * them — but a decoder that declines to answer must not drag the answer down, and two
         * silences must stay silent rather than becoming a number.
         *
         * ⚠ Extracted so it can be tested. The bug it replaces — a `maxHeight` seeded to 1080 and
         * combined with `maxOf`, which could never report anything lower — survived precisely
         * because it lived inside a loop over `MediaCodecList`, a static platform call no unit test
         * reaches. The arithmetic is the part worth pinning; the enumeration is not.
         */
        internal fun tallerOf(
            a: Int?,
            b: Int?,
        ): Int? =
            when {
                a == null -> b
                b == null -> a
                else -> maxOf(a, b)
            }

        /**
         * The tallest picture this decoder reports, or null if it will not say.
         *
         * ⚠ Nullable rather than defaulted, because "the decoder says 1080" and "the decoder would
         * not answer" are different facts and the caller resolves them differently. Collapsing them
         * to a bare 1080 was how a failed query became an affirmative capability claim.
         *
         * `videoCapabilities` is @Nullable in the platform — it is null for an AUDIO codec, which
         * has no video capabilities. Unreachable from the one call site (guarded by `video/hevc`),
         * but written safely rather than relying on a caller staying correct: this returns null for
         * an audio codec instead of throwing inside a `runCatching` that would silently answer 1080.
         */
        private fun maxHeightFor(caps: MediaCodecInfo.CodecCapabilities): Int? =
            runCatching { caps.videoCapabilities?.supportedHeights?.upper }.getOrNull()
    }
}
